### ISSUE TITLE
Editor State recall between plugin loads

### ISSUE THREAD

**Issue by AlexHarker** [iPlug2/iPlug2#117]
Currently the plugin editor state is valid only during the lifetime of the plug-in and is not reloaded. VST3 has a infrastructure to support recall, but other APIs are less obvious. This needs discussion.

**Comment by AlexHarker** [comment]
@olilarkin - so currently there are hooks to serialize and unserialise editor state (which you can override for custom behaviour - by default just stores GUI size/scale). There's also a size/scale pair of methods which you can draw on to save that (as with (Un)SerializeState and parameters).

Do we think any of the formats should be automatically saving GUI state between sessions (obviously VST3 has a specific place to do this, but I'm unaware of anything like that for other formats)? This issue is open for that reason, but if we think it is the programmer's job to implement this then there is nothing to do. 

It could possibly also be provided as an option, although if it is happening through (Un)SerializeState (where I do it in the SRM plugs) the programmer is likely to override anyway, so that might not be a great idea. The difference for VST3 stuff makes this slightly annoying...

**Comment by rexmartin158** [comment]
> 
> ... by default just stores GUI size/scale 

As I posted #773 it is not doing this by default.   When duplicating a plugin or saving a project in a DAW everything gets copied to the new plugin (all parameters, control positions, preset selected, etc.) _**except**_ for the UI size.  It's excellent - and highly desired behavior - that it's saving/copying all the control settings but what is required to get the host to save the UI size?

>
> 
> Do we think any of the formats should be automatically saving GUI state between sessions? 

IMO, plugin UIs should re-open exactly as closed/saved in the project.  I use several commercial plugins in my studio that do this.  The DAW (or the plugin somehow?) saves and re-opens the UI exactly as sized and set when the project was closed. Can't think of any reason why a user would not want this behavior.

**Comment by olilarkin** [comment]
You can add it easily

https://gist.github.com/olilarkin/02a0618cde48f80e22f329513c044e5c

I think that since we need to maintain compatibility with existing state, it's up to the programmer to add this functionality

**Comment by rexmartin158** [comment]
> 
> 
> You can add it easily
> 
> https://gist.github.com/olilarkin/02a0618cde48f80e22f329513c044e5c
> 
> I think that since we need to maintain compatibility with existing state, it's up to the programmer to add this functionality

Great!  Will give that a go.  Thank you for the info/link.

**Comment by rexmartin158** [comment]
> 
> 
> You can add it easily
> 
> https://gist.github.com/olilarkin/02a0618cde48f80e22f329513c044e5c
> 
> I think that since we need to maintain compatibility with existing state, it's up to the programmer to add this functionality

Yes!  This does exactly what I was asking about - duplicated plugin now shares all settings - including UI size.

This code is simple enough - is there a reason why it is not part of the iPlug2 templates/examples?  In other words, are there situations when "chunks" are **_not_** a good choice?

**Comment by olilarkin** [comment]
Chunks are fine, but when using them you need to think carefully about how
your state might change in the future, and how you deal with it.

**Comment by rexmartin158** [comment]
> 
> Chunks are fine, but when using them you need to think carefully about how your state might change in the future, and how you deal with it.

I think there is an error on line 39 of the linked IPlugEffect.cpp example -  

```
int IPlugEffect::UnserializeState(const IByteChunk& chunk, int startPos)
{
    // see above
    int pos = UnserializeEditorState(chunk, Pos);                   <<<<  "pos" should be "startPos"?
    pos = UnserializeParams(chunk, startPos);

    return pos;
}
```

Spoke too soon.  Now the window size is copying but parameter settings are not (they were before).  I set #define PLUG_DOES_STATE_CHUNKS 1.  Will dig into this some more.

**Comment by olilarkin** [comment]
Yes that’s an error, see the first version. I forgot it’s best to serialise
parameters at the end in case you add another parameter

**Comment by rexmartin158** [comment]
> 
> ... it’s best to serialise parameters at the end in case you add another parameter

Yes, I understand.  Thank you for all the info and code example.

Now that I have a better understanding it's unfortunate, IMO, that the "serialize EditorState" functions weren't included in the SerializeState and UnSerializeState functions in iPlugPluginBase.h by default.  I cannot think of a reason why a DAW user would not want this functionality.  I understand, though, that adding it to those code blocks now would break backward compatibility when revising existing plugins that didn't have it unless carefully addressed/overridden.

As you said, it's best if this is established from the beginning.  Since it would not be a good idea to change iPlugPluginBase.h, please consider adding these "serialize/unserialize EditorState" overrides, as you've shown here in iPlugEffect, to the iPlug2 example projects.  As I commented above, yes, IMO, a plugin should always save its UI state.

**Comment by olilarkin** [comment]
The problem with this approach is that it’s also stored in each preset. I
don’t think that a plug-in ui size should change each time you load a
preset. It needs a bit more investigation.

**Comment by rexmartin158** [comment]
> 
> The problem with this approach is that it’s also stored in each preset. I don’t think that a plug-in ui size should change each time you load a preset. It needs a bit more investigation.

Ohhhhh ... I didn't notice that but thank you for pointing it out!  I just checked in Reaper and, yes, it does in fact save the UI size as part of the preset as you say.  I agree that is not desirable.  I subsequently tested a big name commercial plugin and the resizable UI size was not saved in the preset - or at least it did not load it if it was.  So apparently there is some other mechanism beside chunks for storing the UI size in the DAW?  Perhaps the "SerializeState" blocks could have conditional code that skips over the EditorState section when a preset is being loaded vs. the plugin is being copied (if there is a means to detect that)?  Will look at this some more.

**Comment by AlexHarker** [comment]
It gets more complex in that VST3 has a specific spec for storing controller data but the other formats do not. If you build your own internal preset system you can choose what to save within it, but because of the way iplug2 works saving a preset from the host is still likely to store the GUI size etc. I'd also like this to be less work for plug-in developers, but it's hard to propose something that is likely to work for everyone. 

We can retrofit something, as long as developers can access old behaviours (e.g. with overrides).

**Comment by rexmartin158** [comment]
> 
> 
> ...but because of the way iplug2 works saving a preset from the host is still likely to store the GUI size etc. 

Resizable UI is a nice feature but I'm beginning to see how difficult it is to implement cross-platform and cross-DAWs.  For now I'm just letting the plugin open and copy to its default size (not saving EditorState in the chunk).  As long as the UI closes and re-opens to the same size _during_ a project (which it does) I think most users will be OK with it.

Looking ahead, any possibility my suggestion above could work, i.e., conditional code in the "SerializeState" blocks that skips over the EditorState section when a preset is being saved or loaded but not when the plugin is being copied (if there is any means to detect that)?

**Comment by AlexHarker** [comment]
I'm not aware of means to detect it. I'd have to check exactly what we are doing with surreal machines. We have code to store the default size to file on a machine, but I think we save the GUI size with DAW preset/state, but not with internal presets.

**Comment by TBProAudio** [comment]
I dont think that IPLUG needs something special to support GUI size storing and recalling.
We do it in our plugins. We save GUI related parameter (size, width, height) with the DAW chunk and also restore it from there.
And we implemented our own preset system, just to have parameter restoreing (sic e.g. GUI scale) fully under our control :-)

**Comment by rexmartin158** [comment]
> 
> 
> I dont think that IPLUG needs something special to support GUI size storing and recalling.
> We do it in our plugins. We save GUI related parameter (size, width, height) with the DAW chunk and also restore it from there.
> And we implemented our own preset system, just to have parameter restoreing (sic e.g. GUI scale) fully under our control :-)

I checked some of your plugins and was initially inspired to see that they do save their UI size when the plugin is duplicated across tracks in a DAW.  However, I checked further and found that they are also saving that UI size if the settings are saved as a DAW preset.  I think it might annoy users a bit (or think it's a bug) if the UI size jumps around when clicking through presets (I understand you have your own on-board presets that probably do not do this).

So, IMO, there does not seem to be a solution for this right now directly in iPlug2.  We have a choice of capturing both current UI size and parameter settings (and both get saved in the presets) or capturing just the parameter settings.  Not sure how other developers are doing it.  Plugins from Izotope, Fabfilter, etc., do not save the UI size in the DAW presets but they do copy both current UI size and parameter settings when a plugin is copied from track-to-track.  

I feel there may be an "Ah Ha" moment awaiting that might make this relatively easy - but what is it?

**Comment by TBProAudio** [comment]
Hmm, to which plugin are you refering?
Of course factory presets contain all plugin parameter (including GUI scale). But plugin preset recall skips over GUI scale parameter.
So again nothing special is required to avoid GUI scale jumping while browsing presets. Only solid custom preset system is required. And IPLUG provides everything to implement this easily.

**Comment by rexmartin158** [comment]
> 
> 
> Hmm, to which plugin are you refering?

I quickly tested one of your latest - DSEQ3 - in Reaper.  Drag the UI size and save the plugin settings as a preset (in Reaper, not your on-board presets).  Change the UI size then reload the just-saved preset and UI size changes back to what it was when the preset was saved.  If a user saves many presets in the DAW over time, at different UI sizes, then the UI size will jump around when those presets are later clicked through.

> ...plugin preset recall skips over GUI scale parameter.

Your custom system may do this but iPlug does not.  It loads everything including the UI scale at the time the preset was saved (if you have the Editor State chunk enabled).

>So again nothing special is required to avoid GUI scale jumping while browsing presets. 

Yes, something special currently is required as you said above - a "solid custom preset system is required".  Custom anything can be done but it would be great if it could be a standard function in iPlug.

**Comment by TBProAudio** [comment]
Thank you.
Reaper (or any DAW preset system) is not supported. Correct, it works only with plugin preset menu. BTW thats why did our own preset system :-)

**Comment by rexmartin158** [comment]
> 
> 
> Reaper (or any DAW preset system) is not supported. Correct, it works only with plugin preset menu. BTW thats why did our own preset system :-)

Understood (and you have very nice plugins, BTW).  

I have a lot of experience with users who require "fool proofing".  Even with an on-board preset system they will undoubtedly save presets in the DAW then complain that the UI size keeps changing when they recall them.  I am just brainstorming here and wondering if there is something we can do directly in iPlug to address this.  My current approach is to simply not enable the EditorState chunk and let the plugin open to its default size every time a new instance or copy is created.  But users have complained about that too, so...

**Comment by asedev** [comment]
For those case "user A wants this and user B want this" one usually implements a plugin-specific setting. Then everybody may be able to decide about their preferred behavior. Only bad thing: you need to decide about the "default" setting.
Another more audio-related example for this kind of issue is: "ignore dry/wet mix". Usually, this will be saved in the preset but in lots of cases, one has good reasons to ignore it when skipping through presets.

**Comment by olilarkin** [comment]
I've discussed this with a few plug-in devs and the consensus seems to be that many use their own preset manager, and  store the editor dimensions in the regular plug-in state and simply don't save/recall that when saving/loading presets from your preset manager. This means that as discussed, if a user saves/recalls presets via the host the UI will change size.

- I don't believe that VST3 has a solution for saving the UI dimensions with a session and not with a preset. Vst::IEditController::setState() (what I believed @AlexHarker is referring to in the OP) is called for session state and preset state AFAICT. Not looking at VST2 since it's kind of dead as a format now.

- [AUv2](https://developer.apple.com/documentation/audiotoolbox/1534199-general_audio_unit_properties/kaudiounitproperty_classinfofromdocument?language=objc) and [AUv3](https://developer.apple.com/documentation/audiotoolbox/auaudiounit/1387630-fullstatefordocument?language=objc) do have a way of determining document vs preset state. I wonder if it is well supported. 

- [AAX](https://github.com/iPlug2/iPlug2/blob/c9d2d90e77c28d14aeecaefb505ec7b15793d1d4/IPlug/AAX/IPlugAAX.cpp#L543) does have a notification to tell you the origin of the incoming GetChunk() call.

**Comment by rexmartin158** [comment]
I've tested Izotope's Ozone 9 VST3 plugin in a few hosts including Cubase, Ableton and Reaper.  I dragged the UI size larger than default and made some parameter changes - then saved as a preset in host's preset manager.  I removed and reloaded the plugin then clicked on the preset I just saved.  All the parameters loaded but the current UI size did not change.

Also tested Sonible's resizable VST3 plugins with the same result.  Parameters load from host's preset but current UI size does not change.

So, somehow, there is a means to do this in VST3 (and other formats) with host presets.  Unfortunately, I don't know anyone at these other companies who could give us any hints as to _how_ they are doing this.

Not a big issue but it is intriguing.

**Comment by TBProAudio** [comment]
You need to go very deep into iPlug and make this happen:-)
One thing is to extend the parameter class by adding flags related to preset management. 
Second thing is that you need different functions for storing/recalling parameters to/from DAW and presets ...

**Comment by rexmartin158** [comment]
> 
> 
> You need to go very deep into iPlug and make this happen:-)
> One thing is to extend the parameter class by adding flags related to preset management.
> Second thing is that you need different functions for storing/recalling parameters to/from DAW and presets ...

I also want to point out that the two companies I mentioned in my test make plugins that are capable of "inter-plugin communication" i.e., they can send data back and forth to other instances of their plugins elsewhere in the DAW.  If they can do that then they know when the plugin is being copied vs. loaded for the first time, etc.

I imagine that takes some pretty heavy API stuff that is not cross-platform and maybe even DAW-specific.  I wouldn't want to go down that road just to make the UI open a certain size.  Leave it alone.  :)
